group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorConfig.java

```diff
@@ -297,61 +333,92 @@ public class GroupCoordinatorConfig {
      * Also, when a topic is deleted via the delete-topic request, upon propagated metadata update any group's
      *     committed offsets for that topic will also be deleted without extra retention period.
      */
-    public final long offsetsRetentionMs;
+    public long offsetsRetentionMs() {
+        return config.getInt(GroupCoordinatorConfig.OFFSETS_RETENTION_MINUTES_CONFIG) * 60L * 1000L;
+    }
 
     /**
      * Offset commit will be delayed until all replicas for the offsets topic receive the commit
      * or this timeout is reached
      */
-    public final int offsetCommitTimeoutMs;
+    public int offsetCommitTimeoutMs() {
+        return config.getInt(GroupCoordinatorConfig.OFFSET_COMMIT_TIMEOUT_MS_CONFIG);
+    }
```

Comment:
This is an example of the config which is accessed extremely frequently (there are other like this too). I think that having attributes would be better as it avoid having to look it up in the config every time.

Reply:
This is a good question!

The approach of this PR is to make sure `GroupCoordinatorConfig` can see the latest configs to avoid potential bugs (#16394). However, it has side effect which brings extra cost: "volatile" and "lookup/parse".

- The cost of "volatile" (or other similar sync trick) is required if we make  `GroupCoordinatorConfig` see latest configs. 
- The cost of "lookup/parse" could be eliminated if we do a bit refactor for it. For example, we pass `Supplier<GroupCoordinatorConfig>` instead of `GroupCoordinatorConfig` to the callers. By that changes, we can make `GroupCoordinatorConfig` have all immutable pre-created local attributes. The impl of `Supplier<GroupCoordinatorConfig>` will be generated by `KafkaConfig` and it looks like `AtomicReference::get`. However, the side effect is that the usage will get a little ugly:  "config.numThreads()" -> "config.get().numThreads()"

WDYT? BTW, we had a related discussion in https://issues.apache.org/jira/browse/KAFKA-17001

Reply:
None of the configs in this class are dynamic so I am not sure that it is worth it. We could perhaps mention it in the javadoc of the class. If we introduce a dynamic config, we should indeed not use an attribute for it.

For the context, in KafkaConfig, we always had the distinction between val (static values) and def (dynamic ones). We could do the same here, I suppose.